src/output/tree.rs

```diff
@@ -270,30 +289,34 @@ impl<'module> MonoidalPrinter<'module> {
                 }
             }
             Format::Compute(_expr) => self.compile_value(scope, value),
-            Format::Match(head, branches) => {
-                let head = head.eval(&scope);
-                for (pattern, format) in branches {
+            Format::Match(head, branches) => match value {
+                Value::Branch(index, value) => {
+                    let head = head.eval(&scope);
+                    let (pattern, format) = &branches[*index];
                     if let Some(pattern_scope) = head.matches(scope, pattern) {
                         frag.encat(self.compile_decoded_value(&pattern_scope, value, format));
                         return frag;
                     }
+                    panic!("pattern match failure");
                 }
-                panic!("non-exhaustive patterns");
-            }
-            Format::MatchVariant(head, branches) => {
-                let head = head.eval(&scope);
-                for (pattern, _label, format) in branches {
+                _ => panic!("expected branch value"),
```

Comment:
We should possibly keep up the pattern of `other => panic!("expected branch value, found {other:?}")` to ensure that the panics, though rare, give us a reasonable amount of information to work with.